### ISSUE TITLE
[Bug] make_block_ptr leftover DMA aliases the next head when last dim is unaligned

### ISSUE THREAD

**Issue by sunyi0505** [triton-lang/triton-ascend#2261]
### Describe the bug

On NPU, `tl.make_block_ptr` leftover DMA fetches a **full tile width** and does **not** clip to the logical last-dim remainder.
For packed storage `x[T, H, K]` with `K=60`, tile `BK=64`, and stride `H*K`, leftover columns `60:64` of head 0 are physically **head 1 of the same token**. Those aliased values then enter `tl.dot` and turn a legal `QK=60` into `QK=40060`.
This is measured leftover DMA aliasing, not a broken `boundary_check` lowering.
`shape=(T, K)` + `boundary_check=(0, 1)` **does** zero leftover lanes locally. The O(1) failures are:
1. no `boundary_check`
2. declaring `shape` last-dim as `BK` while storage is still logical `K` (mask treats leftover as in-bounds).

### Environment details

- CANN 9.1 (`/home/cann/cann9.1.0`)
- torch_npu 2.9.0.post4
- Triton-Ascend 3.2.2
- dtype: fp16
- `ASCEND_LAUNCH_BLOCKING=1`

Allocator also prints:
The current CANN and Soc versions require processing for 32 padding size, with memory allocation.

`K=60` fp16 = 120 B, which is not 32-byte aligned (next aligned size is 128 B = 64 elements). Leftover is exactly 4 lanes.

## Minimal repro

[repro_leftover_dma.py](https://github.com/user-attachments/files/32363701/repro_leftover_dma.py)

Packed `q/k` with head 0 filled by `1.0` and head 1 filled by `100.0`.

Legal `tl.dot` on head 0: `K * 1 * 1 = 60`.  
If leftover 4 lanes alias head 1: `60 + 4 * 100 * 100 = 40060`.

```python
"""Repro: Triton-Ascend leftover DMA aliases the next head.

Packed x[T, H, K], K=60, tile BK=64, H=2. Head 0 = 1, head 1 = 100.
make_block_ptr strides=(H*K, 1); leftover cols 60:64 physically are head 1.
"""
import os
os.environ.setdefault("ASCEND_LAUNCH_BLOCKING", "1")

import torch
import torch_npu  # noqa: F401
import triton
import triton.language as tl

T, H, K = 64, 2, 60
BT, BK = 64, 64
HEAD0, HEAD1 = 1.0, 100.0

@triton.jit
def cube_qk(q_ptr, k_ptr, out_ptr, T: tl.constexpr, H: tl.constexpr, K: tl.constexpr,
            BT: tl.constexpr, BK: tl.constexpr, USE_BC: tl.constexpr, SHAPE_BK: tl.constexpr):
    kn = BK if SHAPE_BK else K
    p_q = tl.make_block_ptr(q_ptr, (T, kn), (H * K, 1), (0, 0), (BT, BK), (1, 0))
    p_k = tl.make_block_ptr(k_ptr, (kn, T), (1, H * K), (0, 0), (BK, BT), (0, 1))
    b_q = tl.load(p_q, boundary_check=(0, 1)) if USE_BC else tl.load(p_q)
    b_k = tl.load(p_k, boundary_check=(0, 1)) if USE_BC else tl.load(p_k)
    b_a = tl.dot(b_q + 0.0, b_k, allow_tf32=False)
    p_o = tl.make_block_ptr(out_ptr, (BT, BT), (BT, 1), (0, 0), (BT, BT), (1, 0))
    tl.store(p_o, b_a.to(p_o.dtype.element_ty))

@triton.jit
def dump_load(src_ptr, dump_ptr, T: tl.constexpr, H: tl.constexpr, K: tl.constexpr,
              BT: tl.constexpr, BK: tl.constexpr, USE_BC: tl.constexpr, SHAPE_BK: tl.constexpr):
    kn = BK if SHAPE_BK else K
    p = tl.make_block_ptr(src_ptr, (T, kn), (H * K, 1), (0, 0), (BT, BK), (1, 0))
    x = tl.load(p, boundary_check=(0, 1)) if USE_BC else tl.load(p)
    p_d = tl.make_block_ptr(dump_ptr, (BT, BK), (BK, 1), (0, 0), (BT, BK), (1, 0))
    tl.store(p_d, x)

def packed(h0, h1, k, dtype):
    x = torch.empty(T, H, k, device="npu", dtype=dtype)
    x[:, 0, :] = h0
    x[:, 1, :] = h1
    return x.contiguous()

def run_qk(use_bc, shape_bk, dtype):
    q = packed(HEAD0, HEAD1, K, dtype)
    k = packed(HEAD0, HEAD1, K, dtype)
    out = torch.empty(BT, BT, device="npu", dtype=torch.float32)
    cube_qk[(1,)](q, k, out, T=T, H=H, K=K, BT=BT, BK=BK,
                  USE_BC=use_bc, SHAPE_BK=shape_bk, multibuffer=False)
    torch.npu.synchronize()
    return float(out.float()[0, 0].cpu())

def run_dump(use_bc, shape_bk, dtype):
    src = packed(HEAD0, HEAD1, K, dtype)
    dump = torch.full((BT, BK), -1.0, device="npu", dtype=dtype)
    dump_load[(1,)](src, dump, T=T, H=H, K=K, BT=BT, BK=BK,
                    USE_BC=use_bc, SHAPE_BK=shape_bk, multibuffer=False)
    torch.npu.synchronize()
    lanes = dump.float().cpu()[0, K:BK]
    return lanes.tolist(), float(lanes.max())

if __name__ == "__main__":
    dtype = torch.float16
    print(f"valid_dot={K * HEAD0 * HEAD0} alias_dot={K * HEAD0 * HEAD0 + (BK - K) * HEAD1 * HEAD1}")
    for use_bc, shape_bk, name in [
        (True,  False, "bc=1 shape=K"),
        (False, False, "bc=0 shape=K"),
        (True,  True,  "bc=1 shape=BK"),
    ]:
        qk = run_qk(use_bc, shape_bk, dtype)
        lanes, mx = run_dump(use_bc, shape_bk, dtype)
        print(f"{name:16s}  QK={qk:.4g}  leftover={lanes} max={mx:.4g}")
```

|shape last-dim | boundary_check | leftover dump [60:64] | tl.dot A[0,0] | notes|
| --------------- | ------------------- | ----------------------- | ------------- | ----- |
| K=60 | on | [0, 0, 0, 0] | 60 | mask works; leftover does not enter Cube|
| K=60 | off | [100, 100, 100, 100] | 40060 | raw DMA reads through into head 1 |
| BK=64 | on | [100, 100, 100, 100] | 40060 | shape lie; mask treats leftover as valid |

Raw log:
```
K=60 BK=64 leftover=4  valid_dot=60.0 alias_dot=40060.0
cube_qk  bc=1 shape=K   A[0,0]=60
dump     bc=1 shape=K   leftover[0]=[0.0, 0.0, 0.0, 0.0] max=0
cube_qk  bc=0 shape=K   A[0,0]=4.006e+04
dump     bc=0 shape=K   leftover[0]=[100.0, 100.0, 100.0, 100.0] max=100
cube_qk  bc=1 shape=BK  A[0,0]=4.006e+04
dump     bc=1 shape=K   leftover[0]=[100.0, 100.0, 100.0, 100.0] max=100
```
Extra check (not in the snippet above): load unaligned A, then load aligned B into a neighboring UB tile. B was unchanged. So this is source aliasing on leftover DMA, not adjacent-tile UB clobber.

Expected:

  - Leftover DMA should be clipped to the logical remainder (K=60), not fetched at full BK=64.
  - With honest shape=(T, K) and no boundary_check, leftover lanes should still not silently contain the next   row/head. Zero or poison is acceptable; aliasing packed neighbors is not.
  - If hardware DMA must stay 32-byte / tile-width, the compiler should still not present those extra lanes as valid tensor data. Relying on a later Vector boundary_check is too easy to miss, and shape=BK (a common GPU tile pattern) disables that mask.
   
On GPU, OOB without boundary_check is UB, but it does not deterministically splice the next packed head into Cube. On NPU this is a silent O(1) correctness bug: full-tensor RMSE can stay < 1e-3 while max |ref - tri| is O(1) once leftover enters tl.dot.

**Comment by hongziqi** [comment]
Thanks for your feedback. We will arrange relevant engineers to investigate this issue later.

**Comment by hongziqi** [comment]
Thanks for the detailed repro. We reproduced it on Triton-Ascend 3.2, and the same script on A100 gives the same numbers. This matches Triton `tl.make_block_ptr` semantics, not an NPU-specific DMA / IR bug:
* Boundary_check: Use shape as the boundary, with 0 (or other padding) outside the boundary
* No BC: Assuming that the entire block (BT, BK) is valid, retrieve it as a full block, and the compiler will not cut it to K

<img width="677" height="518" alt="Image" src="https://github.com/user-attachments/assets/fe8bc0e0-e48f-41b5-a243-f86c1f756fbb" />

**Comment by sunyi0505** [comment]
I also reproduced it on H100. It is indeed an issue with triton itself. Thanks for your reply.
